### PR TITLE
Feature/create place tests

### PR DESCRIPTION
**Pull request by dimigor**
## JIRA

* [Main JIRA ticket](https://ssu-jira.softserveinc.com/browse/PYTHON-194)

## Code reviewers

- [ ] @meelros
- [ ] @BoroviyOrest
- [ ] @Kardan1123
- [ ] @mizin4ik
- [x] @PetrushynskyiOleksii
- [x] @Sasha1152

### Second Level Review

- [ ] @lhalam

## Summary of change

- Created tests for Place model methods.
- Created tests for Place view.
- Changed response value to constants from responsehelpers module in PlaceView.

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: way_to_home/place/views.py [hunk 1]
```diff
@@ -13,8 +21,6 @@ class PlaceView(View):
     def post(self, request, place_id=None):
         """Handle the request to create a new place object."""
         data = request.body
-        if not data:
-            return HttpResponse('empty json received', status=400)
```
Comment (PetrushynskyiOleksii) [inline way_to_home/place/views.py:17]: Why did you remove it? The `reponsehelpers` module has the corresponding constant for it: `RESPONSE_400_EMPTY_JSON`.

File: way_to_home/place/views.py [hunk 2]
```diff
@@ -45,34 +51,32 @@ def get(self, request, place_id=None):
...
         user = request.user
-
-        if not place_id:
-            return HttpResponse('object id is not received', status=400)
-
         data = request.body
+
         if not data:
-            return HttpResponse('empty json received', status=400)
+            return RESPONSE_400_INVALID_DATA
```
Comment (PetrushynskyiOleksii) [inline way_to_home/place/views.py]: The `responsehelpers` module has `RESPONSE_400_EMPTY_JSON` for such case.

File: way_to_home/tests/unittests/place/testmodel.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+            address='Широка 34, 79052',
+            name='Дім',
+            stop_id=None,
+            user=custom_user
+        )
+
+    def test_get_by_id(self):
+        """Provide tests for `get_by_id` method of certain Place instance."""
+        expected_route = Place.objects.get(id=self.place.id)
+        actual_route = Place.get_by_id(obj_id=self.place.id)
```
Comment (PetrushynskyiOleksii) [inline way_to_home/tests/unittests/place/testmodel.py]: I think it must be `place` instead of `route` :)

File: way_to_home/tests/unittests/place/testmodel.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+        expected_string = '49.842601, 23.968448'
+        actual_string = place.__str__()
+
+        self.assertEqual(expected_string, actual_string)
+
+    def test_create(self):
+        """Provide tests for `create` method of Place model."""
+        user = CustomUser.objects.get(id=2)
+        is_created = Place.create(
```
Comment (PetrushynskyiOleksii) [inline way_to_home/tests/unittests/place/testmodel.py]:
```suggestion
         is_created = Place.create(
```
The value is expected from this name: `True` or `False`. So I think you may change it.

File: way_to_home/tests/unittests/place/testmodel.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            latitude=23.968448,
+            name='Дім',
+            address='Широка 34, 79052',
+            user=user
+        )
+
+        self.assertIsInstance(is_created, Place)
+        self.assertIsNotNone(Place.objects.get(id=self.place.id))
+
+        is_created = Place.create(longitude=Decimal(), latitude=Decimal(), name='', user=CustomUser())
```
Comment (PetrushynskyiOleksii) [inline way_to_home/tests/unittests/place/testmodel.py]:
```suggestion
         is_created = Place.create(longitude=Decimal(), latitude=Decimal(), name='', user=CustomUser())
```
The same situation as mentioned above, change the name of variable, please.

File: way_to_home/tests/unittests/place/testmodel.py [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    """TestCase for providing Place model testing"""
+    def setUp(self):
+        """Method that provides preparation before testing Place model features."""
+        custom_user = CustomUser.objects.create(id=2, email='mymail@icloud.com', is_active=True)
+        custom_user.set_password('qwerty12345')
+        custom_user.save()
+
+        self.client = Client()
+        login = self.client.login(email='mymail@icloud.com', password='qwerty12345')
+        self.assertTrue(login)
```
Comment (PetrushynskyiOleksii) [inline way_to_home/tests/unittests/place/testmodel.py]: You don't need to test Django methods. I think also it is not good to use asserts in `setUp`.

File: way_to_home/tests/unittests/place/testview.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+
+    def setUp(self):
+        """Method that provides preparation before testing Place view`s features."""
+        custom_user = CustomUser.objects.create(id=2, email='mymail@icloud.com', is_active=True)
+        custom_user.set_password('qwerty12345')
+        custom_user.save()
+
+        self.client = Client()
+        login = self.client.login(email='mymail@icloud.com', password='qwerty12345')
+        self.assertTrue(login)
```
Comment (PetrushynskyiOleksii) [inline way_to_home/tests/unittests/place/testview.py]: The same situation as mentioned in the model's test.

File: way_to_home/tests/unittests/place/testview.py [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+        url = reverse('place', args=[])
+        response = self.client.post(url, json.dumps(data), content_type='application/json')
+        response_dict = json.loads(response.content)
+        self.assertEqual(response.status_code, 201)
+        self.assertEqual(response_dict['user_id'], 2)
+        self.assertEqual(response_dict['longitude'], 12.842601)
+        self.assertEqual(response_dict['latitude'], 23.968448)
+        self.assertEqual(response_dict['name'], 'Дім')
+        self.assertEqual(response_dict['address'], 'Суха 3, 79052')
+        self.assertEqual(response_dict['stop_id'], None)
```
Comment (PetrushynskyiOleksii) [inline way_to_home/tests/unittests/place/testview.py]: Did you try to use `assertDictEqual()` instead of this. Or is it not worked?

**Review by PetrushynskyiOleksii — CHANGES_REQUESTED** [review]

File: way_to_home/place/views.py (hunk 1 above)
Comment (dimigor) [reply way_to_home/place/views.py:17]: add this check

File: way_to_home/tests/unittests/place/testmodel.py (hunk 3 above)
Comment (dimigor) [reply way_to_home/tests/unittests/place/testmodel.py]: Yeap :)

File: way_to_home/tests/unittests/place/testmodel.py (hunk 4 above)
Comment (dimigor) [reply way_to_home/tests/unittests/place/testmodel.py]: edited

File: way_to_home/tests/unittests/place/testmodel.py (hunk 5 above)
Comment (dimigor) [reply way_to_home/tests/unittests/place/testmodel.py]: also edited

File: way_to_home/tests/unittests/place/testmodel.py (hunk 6 above)
Comment (dimigor) [reply way_to_home/tests/unittests/place/testmodel.py]: Removed

File: way_to_home/tests/unittests/place/testview.py (hunk 7 above)
Comment (dimigor) [reply way_to_home/tests/unittests/place/testview.py]: Also removed

File: way_to_home/tests/unittests/place/testview.py (hunk 8 above)
Comment (dimigor) [reply way_to_home/tests/unittests/place/testview.py]: Changed to `assertDictEqual()` It's working

**Review by PetrushynskyiOleksii — DISMISSED** [review]

File: way_to_home/tests/unittests/place/testmodel.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+class PlaceViewTest(TestCase):
+    """TestCase for providing Place model testing"""
+    def setUp(self):
+        """Method that provides preparation before testing Place model features."""
+        custom_user = CustomUser.objects.create(id=2, email='mymail@icloud.com', is_active=True)
+        custom_user.set_password('qwerty12345')
+        custom_user.save()
+
+        self.client = Client()
+        login = self.client.login(email='mymail@icloud.com', password='qwerty12345')
```
Comment (lhalam) [inline way_to_home/tests/unittests/place/testmodel.py]: is this variable needed?

File: way_to_home/tests/unittests/place/testview.py [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+    """TestCase for providing Place view testing"""
+
+    def setUp(self):
+        """Method that provides preparation before testing Place view`s features."""
+        custom_user = CustomUser.objects.create(id=2, email='mymail@icloud.com', is_active=True)
+        custom_user.set_password('qwerty12345')
+        custom_user.save()
+
+        self.client = Client()
+        login = self.client.login(email='mymail@icloud.com', password='qwerty12345')
```
Comment (lhalam) [inline way_to_home/tests/unittests/place/testview.py]: is this variable needed?

**Review by lhalam — DISMISSED** [review]

File: way_to_home/tests/unittests/place/testmodel.py (hunk 9 above)
Comment (dimigor) [reply way_to_home/tests/unittests/place/testmodel.py]: removed

File: way_to_home/tests/unittests/place/testview.py (hunk 10 above)
Comment (dimigor) [reply way_to_home/tests/unittests/place/testview.py]: removed

File: way_to_home/tests/unittests/place/testview.py [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+        self.assertEqual(response.status_code, 400)
+
+    def test_put_invalid_data(self):
+        """Method that tests unsuccessful put request with invalid data."""
+
+        data = {
+            'stop_id': "asd",
+        }
+        url = reverse('place', kwargs={'place_id': self.place.id})
+        response = self.client.post(url, json.dumps(data), content_type='application/json')
```
Comment (PetrushynskyiOleksii) [inline way_to_home/tests/unittests/place/testview.py]:
```suggestion
        response = self.client.post(url, json.dumps(data), content_type='application/json')
```
I think it should be `put`

**Review by PetrushynskyiOleksii — CHANGES_REQUESTED** [review]

File: way_to_home/tests/unittests/place/testmodel.py [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+"""
+from decimal import Decimal
+
+from custom_user.models import CustomUser
+from django.test import TestCase, Client
+
+from place.models import Place
+
+
+class PlaceViewTest(TestCase):
```
Comment (Sasha1152) [inline way_to_home/tests/unittests/place/testmodel.py]: maybe PlaceModelTest...

Comment (dimigor) [reply way_to_home/tests/unittests/place/testmodel.py]: Thanks

File: way_to_home/tests/unittests/place/testview.py (hunk 11 above)
Comment (dimigor) [reply way_to_home/tests/unittests/place/testview.py]: Right!

**Review by PetrushynskyiOleksii — DISMISSED** [review]

File: way_to_home/tests/unittests/place/testmodel.py [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+    """TestCase for providing Place model testing"""
+    def setUp(self):
+        """Method that provides preparation before testing Place model features."""
+        custom_user = CustomUser.objects.create(id=2, email='mymail@icloud.com', is_active=True)
+        custom_user.set_password('qwerty12345')
```
Comment (Sasha1152) [inline way_to_home/tests/unittests/place/testmodel.py:18]: Why do you use set_password() here? It does CustomUser.create() method.

Comment (PetrushynskyiOleksii) [reply way_to_home/tests/unittests/place/testmodel.py:18]: I think is a bad idea to use here custom methods. Because if you change the logic of your custom `create` method, u need to change this test also. And what if you remove this method at all?

File: way_to_home/tests/unittests/place/testmodel.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+class PlaceModelTest(TestCase):
+    """TestCase for providing Place model testing"""
+    def setUp(self):
+        """Method that provides preparation before testing Place model features."""
+        custom_user = CustomUser.objects.create(id=2, email='mymail@icloud.com', is_active=True)
+        custom_user.set_password('qwerty12345')
+        custom_user.save()
+
+        self.client = Client()
+        self.client.login(email='mymail@icloud.com', password='qwerty12345')
```
Comment (Sasha1152) [inline way_to_home/tests/unittests/place/testmodel.py]: You don't use `client` instance in this module

Comment (dimigor) [reply way_to_home/tests/unittests/place/testmodel.py]: right

**Review by Sasha1152 — DISMISSED** [review]

File: way_to_home/tests/unittests/place/testmodel.py (hunk 13 above)
Comment (dimigor) [reply way_to_home/tests/unittests/place/testmodel.py:18]: > Why do you use set_password() here? It does CustomUser.create() method.

When I pass a password to create method, User is not created. You know how to fix this?

Comment (dimigor) [reply way_to_home/tests/unittests/place/testmodel.py:18]: > I think is a bad idea to use here custom methods. Because if you change the logic of your custom `create` method, u need to change this test also. And what if you remove this method at all?

Do you know a different way to do that?

Comment (PetrushynskyiOleksii) [reply way_to_home/tests/unittests/place/testmodel.py:18]: For example:
```
user = CustomUser(id=100, email='testuser@mail.com', is_active=True)
user.set_password('testpassword')
user.save()
```
But your way is also applicable. I think it is bad to only use our custom methods here.

Comment (PetrushynskyiOleksii) [reply way_to_home/tests/unittests/place/testmodel.py:18]: > > Why do you use set_password() here? It does CustomUser.create() method.
> 
> When I pass a password to create method, User is not created. You know how to fix this?

The user will be created, but his password will not be hashed.

**Review by PetrushynskyiOleksii — APPROVED** [review]

File: way_to_home/place/views.py [hunk 15]
```diff
@@ -1,9 +1,18 @@
...
 
 from utils.validators import place_data_validator
+from utils.responsehelper import (RESPONSE_200_UPDATED,
+                                  RESPONSE_200_DELETED,
+                                  RESPONSE_400_INVALID_DATA,
+                                  RESPONSE_400_DB_OPERATION_FAILED,
+                                  RESPONSE_400_OBJECT_NOT_RECEIVED,
+                                  RESPONSE_400_EMPTY_JSON,
+                                  RESPONSE_403_ACCESS_DENIED,
+                                  RESPONSE_404_OBJECT_NOT_FOUND,
```
Comment (meelros) [inline way_to_home/place/views.py:14]: remove trailing comma

**Review by meelros — APPROVED** [review]
